components/cloud/azure/src/azblob.rs

```diff
@@ -0,0 +1,740 @@
+    /// retry the entire upload.
+    async fn upload(&self, data: &[u8]) -> Result<(), RequestError> {
+        match timeout(Self::get_timeout(), async {
+            self.client_builder
+                .get_client()
+                .await
+                .map_err(|e| e.to_string())?
+                .as_blob_client(&self.name)
+                .put_block_blob(data.to_vec())
+                .access_tier(self.storage_class)
+                .execute()
+                .await?;
+            Ok(())
+        })
+        .await
+        {
+            Ok(res) => match res {
+                Ok(_) => Ok(()),
+                Err(err) => Err(RequestError::InvalidInput(
+                    err,
+                    "upload block failed".to_owned(),
+                )),
+            },
+            Err(_) => Err(RequestError::TimeOut(
+                "timeout after 15mins for complete in azure storage".to_owned(),
+            )),
+        }
+    }
+
+    fn get_timeout() -> Duration {
+        CONNECTION_TIMEOUT
+    }
+}
+
+// if use azure ad to access the azure blob,
+// it need to update the token at regular intervals,
+// so wrap the client builder
+#[async_trait]
+trait ContainerBuilder: 'static + Send + Sync {
+    async fn get_client(&self) -> io::Result<Arc<ContainerClient>>;
+}
+
+struct SharedKeyContainerBuilder {
+    container_client: Arc<ContainerClient>,
+}
+
+#[async_trait]
+impl ContainerBuilder for SharedKeyContainerBuilder {
+    async fn get_client(&self) -> io::Result<Arc<ContainerClient>> {
+        Ok(self.container_client.clone())
+    }
+}
+
+type TokenCacheType = Arc<RwLock<Option<(TokenResponse, Arc<ContainerClient>)>>>;
+struct TokenCredContainerBuilder {
+    account_name: String,
+    container_name: String,
+    token_resource: String,
+    token_cred: Arc<ClientSecretCredential>,
+    token_cache: TokenCacheType,
+
+    modify_place: Arc<Mutex<bool>>,
+}
+
+impl TokenCredContainerBuilder {
+    fn new(
+        account_name: String,
+        container_name: String,
+        token_resource: String,
+        token_cred: Arc<ClientSecretCredential>,
+    ) -> Self {
+        Self {
+            account_name,
+            container_name,
+            token_resource,
+            token_cred,
+            token_cache: Arc::new(RwLock::new(None)),
+
+            modify_place: Arc::new(Mutex::new(true)),
+        }
+    }
+}
+
+#[async_trait]
+impl ContainerBuilder for TokenCredContainerBuilder {
+    async fn get_client(&self) -> io::Result<Arc<ContainerClient>> {
+        // only the thread get the modify_lock can update the token,
+        // so that this thread can get the token before lock wirte-lock
+        // avoid to block other threads too much time.
+        let mut modify_lock = None;
+        {
+            let token_response = self.token_cache.read().unwrap();
+            if let Some(ref t) = *token_response {
+                let interval = t.0.expires_on - Utc::now();
```

Comment:
is the expires_on use the utc in any place? what if create azure client in Asia🤔

Reply:
Fortunately, the expiration time from the server is the time interval, and the SDK also uses `UTC:: now()`. 
https://github.com/Azure/azure-sdk-for-rust/blob/c5b5e9d86ba8a14586faefaaf8e3cc155c73a24c/sdk/identity/src/token_credentials/client_secret_credentials.rs#L163-L167